node/lib/util/write_repo_ast_util.js

```diff
@@ -81,181 +64,247 @@ const doExec = co.wrap(function *(command) {
-            else if ("string" === typeof change) {
-                // A string indicates files data.
+/**
+ * Return the tree and a map of associated subtrees corresponding to the
+ * specified `flatChanges` in the specified `repo`, and based on the optionally
+ * specified `parent`.  Use the specified `shaMap` to resolve logical shas to
+ * actual written shas (such as for submodule heads).  Use the specified `db`
+ * to write objects.  If the specified `root` is true, generate a `.gitmodules`
+ * file containing an entry for each submodule.
+ *
+ * @async
+ * @param {Boolean}               root
+ * @param {NodeGit.Repository}    repo
+ * @param {NodeGit.Odb}           db
+ * @param {Object}                shaMap maps logical to physical ID
+ * @param {Object}                flatChanges map of changes
+ * @param {Object}                [parent]
+ * @param {Object}                parent.subtrees   filename to return value
+ * @param {NodeGit.Tree}          parent.tree       generated tree object
+ * @param {Object}                parent.submodules name to url
+ * @return {Object}
+ * @return {Object}       return.submodules
+ * @return {Object}       return.subtrees
+ * @return {NodeGit.Tree} return.tree
+ */
+const writeTree = co.wrap(function *(root,
+                                     repo,
+                                     db,
+                                     shaMap,
+                                     changes,
+                                     parent) {
+    let subtrees = {};          // will contain written subtree entries
+    let parentTree = null;      // base root
+    const submodules = {};      // all submodule entries, including children
+
+    // If `parent` is provided, copy values into the above structures.
+
+    if (undefined !== parent) {
+        Object.assign(subtrees, parent.subtrees);
+        parentTree = parent.tree;
+        Object.assign(submodules, parent.submodules);
+    }
 
-                const id = yield hashObject(repo, change);
-                addToTree("100644", "blob", id, path);
-            }
-            else {
-                // If it's not a submodule or a file, it must be a subtree.
+    const wereSubs = 0 !== Object.keys(submodules).length;
 
-                const subTreeId = yield writeHierarchy(change);
-                addToTree("040000", "tree", subTreeId, path);
-            }
-        }
+    // Initialize the builder with the previous tree from which the changes are
+    // derived.
+
+    const builder = yield NodeGit.Treebuilder.create(repo, parentTree);
+
+    const actions = [];  // tree updates to be applied in parallel
 
-        // If no data, make an empty tree
-        if ("" === treeData ) {
-            const builder = yield NodeGit.Treebuilder.create(repo, null);
-            const treeObj = builder.write();
-            return treeObj.tostrS();                                  // RETURN
+    const writeSubtree = co.wrap(function *(filename, entry) {
```

Comment:
The outer function is called `writeTree` and this one will never be called on the outermost tree, so I think it makes some sense.